### PR TITLE
prompt moves based on hint

### PR DESCRIPTION
**Pull request by elferherrera**
The prompt moves and gives space to hint when it is located at the bottom. 

image
the prompt moves upwards based on the suggestion
image

Note: This also removes the issue when the suggestion was being redrawn every second

### CODE REVIEW

File: src/painter.rs [hunk 1]
```diff
@@ -35,6 +35,54 @@ impl PromptCoordinates {
...
+            #[allow(clippy::single_char_pattern)]
+            after_cursor.split("\n").collect::<Vec<&str>>()
+        };
+
+        let hint_lines = if cfg!(windows) {
+            hint.split("\r\n").collect::<Vec<&str>>()
+        } else {
+            #[allow(clippy::single_char_pattern)]
+            hint.split("\n").collect::<Vec<&str>>()
+        };
```
Comment (sholderbach) [inline src/painter.rs:72]: For better readability how about making the switch in the definition of a single `const NEW_LINE_SEPARATOR: &str` ?

File: src/engine.rs [hunk 2]
```diff
@@ -978,7 +980,25 @@ impl Reedline {
...
+                hint,
+                self.use_ansi_coloring,
+            )?;
+        } else {
+            self.painter.queue_buffer(
+                &lines.before_cursor_lines,
+                &lines.after_cursor_lines,
+                &hint,
+            )?;
+        }
```
Comment (sholderbach) [inline src/engine.rs:1000]: This logic is probably also required in the full_repaint and history paint cases. Can we lift that into `Painter::queue_buffer`?

File: src/engine.rs [hunk 3]
```diff
@@ -978,7 +980,25 @@ impl Reedline {
     fn buffer_paint(&mut self, prompt: &dyn Prompt) -> Result<()> {
         let (highlighted_line, hint) = self.prepare_buffer_content(prompt);
 
-        self.painter.queue_buffer(highlighted_line, hint)?;
+        let lines = PromptLines::from_strings(&highlighted_line.0, &highlighted_line.1, &hint);
```
Comment (sholderbach) [inline src/engine.rs:983]: How about letting `prepare_buffer_content()` directly returning the prompt lines?

File: src/painter.rs [hunk 4]
```diff
@@ -35,6 +35,54 @@ impl PromptCoordinates {
...
+
+        Self {
+            before_cursor_lines,
+            after_cursor_lines,
+            hint_lines,
+        }
+    }
+
+    pub fn required_lines(&self) -> u16 {
+        (self.before_cursor_lines.len() + self.hint_lines.len()) as u16
```
Comment (sholderbach) [inline src/painter.rs:82]: Is this exhaustive if there are multiple lines after the cursor?

File: src/painter.rs (hunk 1 above)
Comment (sholderbach) [reply src/painter.rs:72]: Alternatively... is `str::lines` doing the right thing already?

File: src/painter.rs [hunk 5]
```diff
@@ -102,47 +155,38 @@ impl Painter {
...
-            #[allow(clippy::single_char_pattern)]
-            after_cursor.split("\n")
-        };
-
-        let mut commands = self.stdout.queue(MoveTo(
+    pub fn queue_buffer(
+        &mut self,
+        before_cursor_lines: &[&str],
+        after_cursor_lines: &[&str],
+        hint: &str,
```
Comment (sholderbach) [inline src/painter.rs:162]: Can this simply be `PromptLines`?

File: src/painter.rs [hunk 6]
```diff
@@ -194,6 +238,21 @@ impl Painter {
         hint: String,
```
Comment (sholderbach) [inline src/painter.rs:238]: Also simplification via `PromptLines`

File: src/painter.rs [hunk 7]
```diff
@@ -35,6 +35,54 @@ impl PromptCoordinates {
     }
 }
 
+pub struct PromptLines<'prompt> {
```
Comment (sholderbach) [inline src/painter.rs:38]: Bikeshed: `BufferView` as it does not care about the `Prompt` directly and `View` may reflect the reference nature. Same for the lifetime name

**Review by sholderbach — COMMENTED** [review]
Awesome that you are tackling that! I like the simplification. Will try to give it a thorough testing

**Comment by jntrnr** [comment]
Seeming some places where we need to output /r/n to keep things aligned:

image

**Comment by elferherrera** [comment]
@jntrnr can you check with the new commit?

**Comment by jntrnr** [comment]
@elferherrera same issue -

image

File: src/painter.rs [hunk 8]
```diff
@@ -15,6 +17,8 @@ use {
     unicode_width::UnicodeWidthStr,
 };
 
+const END_LINE: &str = if cfg!(windows) { "\r\n" } else { "n" };
```
Comment (sholderbach) [inline src/painter.rs:20]: Oopsie

**Comment by elferherrera** [comment]
@jntrnr @sholderbach @fdncred do you mind checking the new commit?

**Comment by jntrnr** [comment]
@elferherrera - same issue

image

**Comment by fdncred** [comment]
ya, same for me too. 
Screen Shot 2021-12-29 at 4 16 14 PM

**Comment by elferherrera** [comment]
@fdncred @sholderbach do you mind checking again this commit?

**Comment by sholderbach** [comment]
Looking good to me. Even wrapping lines seem to be good at the bottom of the screen. The only thing i noticed that if the last line is reached with actual content, it jumps up a line to leave a blank line after a full repaint.

**Comment by fdncred** [comment]
I'd land this one and continue to work on other issues like 'very long buffer' and 'jumps up a line'.

**Comment by elferherrera** [comment]
@fdncred I think that is a good idea. I dont want to make the PR larger

**Comment by jntrnr** [comment]
Tried out the latest - it doesn't seem to let you work with the last line of the screen:
image

I also noticed that if you have a multiline history hint at the bottom, each character scrolls:
image

**Comment by jntrnr** [comment]
Here's an animation to see these a little bit better:
Peek 2021-12-31 08-00

**Comment by fdncred** [comment]
it's interesting that I don't get this behavior on Mac.

**Comment by jntrnr** [comment]
@elferherrera - thanks for banging this into shape!

**Pull request merged**